File: mediation/mediatedviews/GoogleAdMob/ANAdAdapterNativeAdMob.m
/*   Copyright 2016 APPNEXUS INC
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

#import "ANAdAdapterNativeAdMob.h"
#import "ANAdAdapterBaseDFP.h"
#import <GoogleMobileAds/GoogleMobileAds.h>
#if __has_include(<AppNexusSDK/AppNexusSDK.h>)
#import <AppNexusSDK/AppNexusSDK.h>
#else
#import "ANProxyViewController.h"
#endif

@interface ANAdAdapterNativeAdMob () <GADNativeAdLoaderDelegate, GADNativeAdDelegate>

@property (nonatomic) GADAdLoader *nativeAdLoader;
@property (nonatomic) ANProxyViewController *proxyViewController;
@property (nonatomic) GADNativeAd *nativeAd;

@end

@implementation ANAdAdapterNativeAdMob

@synthesize requestDelegate = _requestDelegate;
@synthesize nativeAdDelegate = _nativeAdDelegate;
@synthesize expired = _expired;


#pragma mark - ANNativeCustomAdapter

- (instancetype)init {
    if (self = [super init]) {
        self.proxyViewController = [[ANProxyViewController alloc] init];
    }
    return self;
}

- (void)requestNativeAdWithServerParameter:(nullable NSString *)parameterString
                                  adUnitId:(nullable NSString *)adUnitId
                       targetingParameters:(nullable ANTargetingParameters *)targetingParameters {


    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    self.nativeAdLoader = [[GADAdLoader alloc] initWithAdUnitID:adUnitId
                                             rootViewController:(UIViewController *)self.proxyViewController
                                                        adTypes:@[GADAdLoaderAdTypeNative]
                                                        options:@[]];
    self.nativeAdLoader.delegate = self;
    [self.nativeAdLoader loadRequest:[ANAdAdapterBaseDFP googleAdMobRequestFromTargetingParameters:targetingParameters rootViewController:nil]];
}

- (void)registerViewForImpressionTrackingAndClickHandling:(nonnull UIView *)view
                                   withRootViewController:(nonnull UIViewController *)rvc
                                           clickableViews:(nullable NSArray *)clickableViews {
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    self.proxyViewController.rootViewController = rvc;
    self.proxyViewController.adView = view;
    if (self.nativeAd) {
        if ([view isKindOfClass:[GADNativeAdView class]]) {
            GADNativeAdView *nativeContentAdView = (GADNativeAdView *)view;
            [nativeContentAdView setNativeAd:self.nativeAd];
        } else {
            ANLogError(@"Could not register native ad view––expected a view which is a subclass of GADNativeAdView");
        }
        return;
    }
}

#pragma mark - GADAdLoaderDelegate

- (void)adLoader:(GADAdLoader *)adLoader didFailToReceiveAdWithError:(NSError *)error {
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    ANLogError(@"Error loading Google native ad: %@", error);
    [self.requestDelegate didFailToLoadNativeAd:[ANAdAdapterBaseDFP responseCodeFromRequestError:error]];
}

#pragma mark - GADNativeAppInstallAdLoaderDelegate

- (void)adLoader:(GADAdLoader *)adLoader didReceiveNativeAd:(GADNativeAd *)nativeAd
{
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    self.nativeAd = nativeAd;
    ANNativeMediatedAdResponse *response = [[ANNativeMediatedAdResponse alloc] initWithCustomAdapter:self
                                                                                         networkCode:ANNativeAdNetworkCodeAdMob];
    nativeAd.delegate = self;
    response.title = nativeAd.headline;
    response.body = nativeAd.body;
    response.iconImageURL = nativeAd.icon.imageURL;
    response.mainImageURL = ((GADNativeAdImage *)[nativeAd.images firstObject]).imageURL;
    response.callToAction = nativeAd.callToAction;
    response.rating = [[ANNativeAdStarRating alloc] initWithValue:[nativeAd.starRating floatValue]
                                                            scale:5.0];
    response.customElements = @{kANNativeElementObject:nativeAd};
    [self.requestDelegate didLoadNativeAd:response];
}

#pragma mark - GADNativeAdDelegate

- (void)nativeAdWillPresentScreen:(GADNativeAd *)nativeAd {
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    [self.nativeAdDelegate willPresentAd];
    [self.nativeAdDelegate didPresentAd];
}

- (void)nativeAdWillDismissScreen:(GADNativeAd *)nativeAd {
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    [self.nativeAdDelegate willCloseAd];
}

- (void)nativeAdDidDismissScreen:(GADNativeAd *)nativeAd {
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    [self.nativeAdDelegate didCloseAd];
}

- (void)nativeAdWillLeaveApplication:(GADNativeAd *)nativeAd {
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    [self.nativeAdDelegate willLeaveApplication];
}

- (void)nativeAdDidRecordImpression:(GADNativeAd *)nativeAd{
    ANLogTrace(@"%@ %@", NSStringFromClass([self class]), NSStringFromSelector(_cmd));
    [self.nativeAdDelegate adDidLogImpression];
}

@end

File: examples/SimpleMediation/SimpleMediation/ANAdAdapterNativeAdMob.swift
//
//  CustomAdapter.swift
//  SimpleMediation
//
//  Created by System on 17/05/22.
//  Copyright © 2022 Xandr. All rights reserved.
//

import Foundation
import AppNexusSDK
import GoogleMobileAds

@objc(ANAdAdapterNativeAdMob)
public class ANAdAdapterNativeAdMob : NSObject , ANNativeCustomAdapter , GADNativeAdLoaderDelegate, GADNativeAdDelegate {    
    
    
    var nativeAdLoader: GADAdLoader!
    var proxyViewController : ANProxyViewController!
    var nativeAd: GADNativeAd?
    
    override init() {
        super.init()
//        hasExpired = true
        proxyViewController = ANProxyViewController()
    }
    
    public func requestNativeAd(
        withServerParameter parameterString: String?,
        adUnitId: String?,
        targetingParameters: ANTargetingParameters?
    ) {
        nativeAdLoader = GADAdLoader(  adUnitID: adUnitId!,
                                       rootViewController: proxyViewController as? UIViewController,  adTypes: [GADAdLoaderAdType.native],
                                       options: [])
        nativeAdLoader.delegate = self
        nativeAdLoader.load(GADRequest())
        
    }
    
    public var requestDelegate: ANNativeCustomAdapterRequestDelegate?
//    @nonobjc public var hasExpired: Bool?
    public var nativeAdDelegate: ANNativeCustomAdapterAdDelegate?
    public var expired: ObjCBool?

    public func hasExpired() -> DarwinBoolean{
        return false
    }
    public func registerView(forImpressionTrackingAndClickHandling view: UIView, withRootViewController rvc: UIViewController, clickableViews: [Any]?) {
        
        print("registerView by Ab")
    
//    public func registerView(forImpressionTrackingAndClickHandling view: UIView, withRootViewController rvc: UIViewController, clickableViews: [Any]?) {
        proxyViewController.rootViewController = rvc
        proxyViewController.adView = view
        if (nativeAd != nil) {
            if view is GADNativeAdView {
                let nativeContentAdView = view as? GADNativeAdView
                nativeContentAdView?.nativeAd = nativeAd
            }


            return;
        }
    }
    
    public func adLoader(_ adLoader: GADAdLoader, didReceive nativeAd: GADNativeAd) {
//        self.hasExpired = false

        
        let response = ANNativeMediatedAdResponse(
            customAdapter: self,
            networkCode: ANNativeAdNetworkCode.adMob)
        nativeAd.delegate = self
        response?.title = nativeAd.headline
        response?.body = nativeAd.body
        response?.iconImageURL = nativeAd.icon?.imageURL
        response?.iconImageURL = nativeAd.icon?.imageURL
        response?.mainImageURL = (nativeAd.images?.first)?.imageURL
        response?.callToAction = nativeAd.callToAction
        response?.rating = ANNativeAdStarRating(
            value: CGFloat(nativeAd.starRating?.floatValue ?? 0.0),
            scale: Int(5.0))
        response?.customElements = [
            kANNativeElementObject: nativeAd
        ]
        requestDelegate?.didLoadNativeAd!(response!)
    }
    
    public func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: Error) {
        requestDelegate?.didFail!(toLoadNativeAd: ANAdResponseCode.unable_TO_FILL())
        
    }
    
    
    
    public func nativeAdDidRecordImpression(_ nativeAd: GADNativeAd) {
      // The native ad was shown.
        self.nativeAdDelegate?.adDidLogImpression!()
    }

    public func nativeAdDidRecordClick(_ nativeAd: GADNativeAd) {
      // The native ad was clicked on.
        self.nativeAdDelegate?.adWasClicked!()

    }

    public func nativeAdWillPresentScreen(_ nativeAd: GADNativeAd) {
      // The native ad will present a full screen view.
        self.nativeAdDelegate?.didPresentAd!()

    }

    public func nativeAdWillDismissScreen(_ nativeAd: GADNativeAd) {
      // The native ad will dismiss a full screen view.
        self.nativeAdDelegate?.willCloseAd!()

    }

    public func nativeAdDidDismissScreen(_ nativeAd: GADNativeAd) {
      // The native ad did dismiss a full screen view.
        self.nativeAdDelegate?.didCloseAd!()

    }

    public func nativeAdWillLeaveApplication(_ nativeAd: GADNativeAd) {
      // The native ad will cause the application to become inactive and
      // open a new application.
        self.nativeAdDelegate?.willLeaveApplication!()

    }
 
//    public func hasExpired() -> Bool {
//        return true
//    }
 
    
    
}
